pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -0,0 +1,218 @@
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/client/typed/dynamic"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+	"k8s.io/kubernetes/pkg/watch"
+)
+
+const ResourceResyncTime = 30 * time.Second
+
+type GarbageCollector struct {
+	restMapper         meta.RESTMapper
+	clientPool         dynamic.ClientPool
+	dirty_queue        *workqueue.Type
+	scannerStores      []cache.Store
+	scannerControllers []*framework.Controller
+}
+
+func NewGarbageCollector(clientPool dynamic.ClientPool, resources []unversioned.GroupVersionResource) *GarbageCollector {
+	gc := &GarbageCollector{
+		clientPool:  clientPool,
+		dirty_queue: workqueue.New(),
+		// TODO: should use a dynamic RESTMapper built from the discovery results.
+		restMapper: registered.RESTMapper(),
+	}
+	for _, resource := range resources {
+		scannerStore, scannerController := framework.NewInformer(
+			&cache.ListWatch{
+				ListFunc: func(options api.ListOptions) (runtime.Object, error) {
+					client, err := gc.clientPool.ClientForGroupVersion(resource.GroupVersion())
+					if err != nil {
+						return nil, err
+					}
+					// APIResource.Kind is not used by the dynamic client, so
+					// leave it empty. We want to list this resource in all
+					// namespaces if it's namespace scoped, so leave
+					// APIResource.Namespaced as false is all right.
+					apiResource := unversioned.APIResource{Name: resource.Resource}
+					return client.Resource(&apiResource, api.NamespaceAll).UnversionedList(options)
+				},
+				WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+					client, err := gc.clientPool.ClientForGroupVersion(resource.GroupVersion())
+					if err != nil {
+						return nil, err
+					}
+					// APIResource.Kind is not used by the dynamic client, so
+					// leave it empty. We want to list this resource in all
+					// namespaces if it's namespace scoped, so leave
+					// APIResource.Namespaced as false is all right.
+					apiResource := unversioned.APIResource{Name: resource.Resource}
+					return client.Resource(&apiResource, api.NamespaceAll).UnversionedWatch(options)
+				},
+			},
+			// TODO: It's impossible to understand every Kind in the cluster.
+			nil,
+			ResourceResyncTime,
+			framework.ResourceEventHandlerFuncs{},
+		)
+		gc.scannerControllers = append(gc.scannerControllers, scannerController)
+		gc.scannerStores = append(gc.scannerStores, scannerStore)
+	}
+	return gc
+}
+
+type StoreKey struct {
+	Store cache.Store
+	Key   string
+}
+
+func (gc *GarbageCollector) scanner() {
+	for {
+		func() {
+			for _, scannerStore := range gc.scannerStores {
+				keys := scannerStore.ListKeys()
+				// TODO: limit the size of the dirty queue
+				for _, key := range keys {
+					gc.dirty_queue.Add(StoreKey{scannerStore, key})
+				}
+			}
+		}()
+	}
+}
+
+func (gc *GarbageCollector) worker() {
+	for {
+		func() {
```

Comment:
... the for loop and func() can be omitted since you call this from a wait.Until.